src/Symfony/Component/Security/Http/Authenticator/MagicLinkAuthenticator.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\Authenticator;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\InvalidMagicLinkAuthenticationException;
+use Symfony\Component\Security\Http\Authentication\AuthenticationFailureHandlerInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationSuccessHandlerInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;
+use Symfony\Component\Security\Http\HttpUtils;
+use Symfony\Component\Security\Http\MagicLink\Exception\InvalidMagicLinkExceptionInterface;
+use Symfony\Component\Security\Http\MagicLink\MagicLoginLinkerInterface;
+
+/**
+ * @author Ryan Weaver <ryan@symfonycasts.com>
+ */
+class MagicLinkAuthenticator extends AbstractAuthenticator implements InteractiveAuthenticatorInterface
+{
+    private $magicLoginLinkHelper;
+    private $httpUtils;
+    private $successHandler;
+    private $failureHandler;
+    private $options;
+
+    public function __construct(MagicLoginLinkerInterface $magicLoginLinkHelper, HttpUtils $httpUtils, AuthenticationSuccessHandlerInterface $successHandler, AuthenticationFailureHandlerInterface $failureHandler, array $options)
+    {
+        $this->magicLoginLinkHelper = $magicLoginLinkHelper;
+        $this->httpUtils = $httpUtils;
+        $this->successHandler = $successHandler;
+        $this->failureHandler = $failureHandler;
+        $this->options = $options;
+    }
+
+    public function supports(Request $request): ?bool
+    {
+        return $this->httpUtils->checkRequestPath($request, $this->options['check_route']);
+    }
+
+    public function authenticate(Request $request): PassportInterface
+    {
+        $uri = $request->getUri();
+
+        return new SelfValidatingPassport(
+            new UserBadge($uri, function ($uri) {
+                try {
+                    $user = $this->magicLoginLinkHelper
+                        ->consumeMagicLink($uri);
+                } catch (InvalidMagicLinkExceptionInterface $e) {
+                    throw new InvalidMagicLinkAuthenticationException('Magic link could not be validated.', 0, $e);
+                }
+
+                return $user;
+            }),
```

Comment:
> What if - instead of a UserInterface object - we store the username (or identifier) in the token? Then this authenticator can use the user provider to find the user object for this login action

Yes, I think this is a BIG question. I chose to use a foreign key for referential integrity, for the exact reason you mentioned: if the user changes their username, for example. 

But... we could (maybe?) still have a foreign key with your idea:

1) Allow the user to say "Hey! My user is an entity, the class is `App\Entity\User`, and so please handle everything for me". In that case, we could generate the schema with a foreign key to the User entity.

2) Or, allow the user to say "Hey! My user is *not* an entity, so I will create some "storage" service that determines how to store & load the user".

Well, the PR already has 2. So I guess it would just mean that the "default Doctrine" implementation would change to be not-entity based...